feat(Menu): adding kind prop for the items and MenuDivider component

mnajdova opened the pull request:
This PR adds the kind prop in the items prop in the Menu, for creating different components inside the menu (MenuItem, MenuDivider- component added as part of this PR). The PR is introduced in order to fix https://github.com/stardust-ui/react/issues/667 

For future work, we may leverage this kind property for extracting the MenuItem with submenus as different component, a MenuHeader component etc.

Example:
Vertical pointing menu:
```
import React from 'react'
import { Menu } from '@stardust-ui/react'
const items = [
  {
    key: 'editorials',
    content: 'Editorials',
  },
  {
    key: 'review',
    content: 'Reviews',
  },
  {
    key: 'events',
    content: 'Upcoming Events',
  },
]

const MenuExampleVerticalPointing = () => (
  <Menu defaultActiveIndex={0} items={items} vertical pointing />
)

export default MenuExampleVerticalPointing
```

![image](https://user-images.githubusercontent.com/4512430/50834279-ea8f8b80-1353-11e9-8fb5-ac900deff7d0.png)


With separator between the last two elements:
```
import React from 'react'
import { Menu } from '@stardust-ui/react'
const items = [
  {
    key: 'editorials',
    content: 'Editorials',
  },
  {
    key: 'review',
    content: 'Reviews',
  },
  {
    key: 'separator',
    kind: 'MenuSeparator',
  },
  {
    key: 'events',
    content: 'Upcoming Events',
  },
]

const MenuExampleKind = () => (
  <Menu defaultActiveIndex={0} items={items} vertical pointing="start" />
)

export default MenuExampleKind
```

![image](https://user-images.githubusercontent.com/4512430/50834309-f8dda780-1353-11e9-8fee-47bb40cdec99.png)

Commit:
-added menu separator component
-added kind prop for the items inside the menu

Commit:
-removed styles for itemSeparator in the MenuItem styles

Commit:
Merge branch 'master' into feat/menu-kind-prop

Commit:
-updated changelog

Commit:
Merge branch 'master' into feat/menu-kind-prop

miroslavstastny commented on src/themes/teams/components/Menu/menuItemStyles.ts:
```diff
@@ -197,8 +174,6 @@ const menuItemStyles: ComponentSlotStylesInput<MenuItemPropsAndState, MenuVariab
         display: 'flex',
       }),
 
-      ...itemSeparator({ props, variables: v, theme }),
```
I am not sure that we want to remove separators in Teams theme menu

mnajdova replied:
I had the same doubt, but wasn't sure how can I show the separator. Will add example with some other menu type that does not have separator by default then... Thanks for the suggestion.

mnajdova replied:
Addressed the comment, added example for some other variants, take a look in the description of the PR.

Commit:
-fixed changelog

Commit:
-added example for using the kind prop
-reverted removed item separator from teams theme

layershifter commented on docs/src/examples/components/Menu/Usage/MenuExampleKind.shorthand.tsx:
```diff
@@ -0,0 +1,15 @@
+const items = [
+  { key: 'editorials', content: 'Editorials' },
+  { key: 'review', content: 'Reviews' },
+  { key: 'separator', kind: 'MenuSeparator' },
```
As I remember we decided to use lowercase without prefixes in the `kind` prop, i.e. `separator`.
And technically, it's a `divider`

layershifter commented on src/components/Menu/MenuItem.tsx:
```diff
@@ -144,6 +146,7 @@ class MenuItem extends AutoControlledComponent<ReactProps<MenuItemProps>, MenuIt
     defaultMenuOpen: PropTypes.bool,
     onActiveChanged: PropTypes.func,
     inSubmenu: PropTypes.bool,
+    kind: PropTypes.string,
```
```suggestion
    kind: PropTypes.oneOf(['divider']),
```

mnajdova replied:
Why should the type here be `oneOf` containing divider? MenuItem will actually never be created if the kind prop is divider. For the menu item it can be 'item' according to the convention, or nothing as this is the default component. Am I wrong assuming this?

layershifter replied:
So, `kind: PropTypes.oneOf(['divider', 'item']),`? We should disallow to pass invalid values to `kind`

mnajdova replied:
My point was that in the MenuItem, it should be `oneOf(['item'])`, and in the divider should be `oneOf(['divider'])` as the MenuItem will never be created if the kind prop is something other then item (or null, but the prop is anyways optional - anything else provided should throw an error), and the MenuDivider will never be created if the kind prop is something other then the 'divider'. Would this be sufficient?

layershifter replied:
Resolved in personal conversation :)

layershifter commented on src/components/Menu/MenuItem.tsx:
```diff
@@ -105,6 +105,8 @@ export interface MenuItemProps
   /** Indicates whether the menu item is part of submenu. */
   inSubmenu?: boolean
+
+  kind?: string
```
```suggestion
  kind?: 'divider'
```

layershifter commented on src/themes/teams/components/Menu/menuSeparatorStyles.ts:
```diff
@@ -0,0 +1,19 @@
+        : {
+            borderLeft: `1px solid ${p.primary ? v.primaryBorderColor : v.borderColor}`,
+          }),
+    }
```
```js
const borderColor = p.primary ? v.primaryBorderColor : v.borderColor
const borderType = p.vertical ? 'borderTop' : 'borderLeft'

return {
  [borderType]: `1px solid ${borderColor}`
}
```

We can change naming, but this looks to me more readable and compact :)

Commit:
-renamed MenuSeparator to MenuDivider and added test
-fixed types of the kind prop
-improved styles

Commit:
-added typings for the kind prop
-added service for getting the kind prop
-removed kind prop from MenuItem and MenuDivider

Commit:
-removed unused imports

layershifter commented on src/components/Menu/Menu.tsx:
```diff
@@ -10,13 +10,17 @@ import {
+import { ReactProps, ShorthandCollection } from '../../../types/utils'
+import MenuDivider from './MenuDivider'
+
+export type MenuItemKindOptions = 'divider' | 'item'
```
```suggestion
export type MenuItemKinds = 'divider' | 'item'
```

`MenuItemKinds` may be? Or even `MenuShorthandKinds` because they are related to shorthand items, not to `MenuItem`

mnajdova replied:
`MenuShorthandKinds` it is :)

layershifter commented on docs/src/examples/components/Menu/Usage/MenuExampleKind.shorthand.tsx:
```diff
@@ -0,0 +1,15 @@
+  <Menu defaultActiveIndex={0} items={items} vertical pointing="start" />
+)
+
+export default MenuExampleKind
```
```suggestion
export default MenuExampleDivider
```

As I understand this example should show us the usage of `MenuDivider`, SUI uses `Content` section of docs to show elements usage: https://semantic-ui.com/collections/menu.html#header

Commit:
-renamed MenuItemKindProp to MenuShorthandKinds
-renamed Usage examples to Content examples

Commit:
-changed the order of the content and variations menu examples

Commit:
Merge branch 'master' into feat/menu-kind-prop

# Conflicts:
#	CHANGELOG.md
